### PR TITLE
Support NuGet lockfiles (Updated)

### LINKED ISSUE #1303: Support for NuGet package lock files
In NuGet 4.9 I think it is they added support for package lock files using the setting `RestorePackagesWithLockFile` which is part of your `.csproj` or `.props` file chain. This creates a file called `packages.lock.json` alongside your project file. The spec for this can be found at https://github.com/NuGet/Home/wiki/Enable-repeatable-package-restore-using-lock-file

### PR DESCRIPTION
**Pull request by na1307**
Fix #1303

Since #6031 was last updated, Dependabot NuGet logic has been rewritten as C#, so I've put this PR up.

I'm familiar with C#, but not with Ruby. I don't know how to add a test for this. Any advice is welcome.

### CODE REVIEW

**Comment by mheguy-flo** [comment]
The linting failure for fetch_files is because your changes pushed the complexity over the limit.

**Comment by na1307** [comment]
I don't know why the test failed. It was successful when I tested it locally.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/UpdaterWorker.cs [hunk 1]
```diff
@@ -117,6 +122,20 @@ public async Task RunAsync(string repoRootPath, string workspacePath, string dep
...
+        string lockPath)
+    {
+        _logger.Log($"Running for lock file [{Path.GetRelativePath(repoRootPath, lockPath)}]");
+        if (!File.Exists(lockPath))
+        {
+            _logger.Log($"File [{lockPath}] does not exist.");
+            return;
+        }
+
+        await Process.Start("dotnet", $"restore --force-evaluate {Path.GetDirectoryName(lockPath)}").WaitForExitAsync();
```
Comment (hangy) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/UpdaterWorker.cs:136]: Maybe use [`ProcessEx.RunAsync`](https://github.com/dependabot/dependabot-core/blob/de34b53e2e5cac84a4e0f2fd0bdb2276e01e1ca7/nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/ProcessExtensions.cs#L9) so that the process output is redirect to the same stream as the other times where `dotnet` is called externally?

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/UpdaterWorker.cs [hunk 2]
```diff
@@ -39,6 +41,9 @@ public async Task RunAsync(string repoRootPath, string workspacePath, string dep
             case ".vbproj":
                 await RunForProjectAsync(repoRootPath, workspacePath, dependencyName, previousDependencyVersion, newDependencyVersion, isTransitive);
                 break;
+            case ".json":
+                await RunForPackagesLockAsync(repoRootPath, workspacePath);
```
Comment (JoeRobich) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/UpdaterWorker.cs:45]: I don't think we have guarantees about the ordering of dependencies we invoke the native tool with. Meaning this may be invoked prior to updating the project itself.

You could consider walking up from the project file being updated and looking for a package lock to update. You could then invoke restore after the update operation has been completed.

File: nuget/lib/dependabot/nuget/file_updater.rb [hunk 3]
```diff
@@ -25,7 +25,8 @@ def self.updated_files_regex
           /^dotnet-tools\.json$/i,
           /^Directory\.Build\.props$/i,
           /^Directory\.Build\.targets$/i,
-          /^Packages\.props$/i
+          /^Packages\.props$/i,
```
Comment (JoeRobich) [inline nuget/lib/dependabot/nuget/file_updater.rb:28]: I would probably revert these changes and instead identify package lock file that need updating based on the projects being updated in the native tool.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/UpdaterWorker.cs (hunk 2 above)
Comment (na1307) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/UpdaterWorker.cs:45]: I tried that, but in a solution using central package management, it only updates the lock file for one project. Is there a way?

Comment (na1307) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/UpdaterWorker.cs:45]: This is the code I wrote:
```cs
private async Task RunUpdaterAsync(
        string repoRootPath,
        string projectPath,
        string dependencyName,
        string previousDependencyVersion,
        string newDependencyVersion,
        bool isTransitive)
    {
        ...

        // Update lock file if exists
        if (File.Exists(Path.Combine(Path.GetDirectoryName(projectPath), "packages.lock.json")))
        {
            await LockFileUpdater.UpdateLockFileAsync(repoRootPath, projectPath, _logger);
        }
    }
```

```cs
internal static class LockFileUpdater
{
    public static async Task UpdateLockFileAsync(
        string repoRootPath,
        string projectPath,
        Logger logger)
    {
        var lockPath = Path.Combine(Path.GetDirectoryName(projectPath), "packages.lock.json");
        logger.Log($"  Running for lock file [{Path.GetRelativePath(repoRootPath, lockPath)}]");
        if (!File.Exists(lockPath))
        {
            logger.Log($"    File [{Path.GetRelativePath(repoRootPath, lockPath)}] does not exist.");
            return;
        }

        var (exitCode, stdout, stderr) = await ProcessEx.RunAsync("dotnet", $"restore --force-evaluate {projectPath}");
        if (exitCode != 0)
        {
            logger.Log($"    Lock file update failed.\nSTDOUT:\n{stdout}\nSTDERR:\n{stderr}");
            return;
        }

        logger.Log($"    Saved [{Path.GetRelativePath(repoRootPath, lockPath)}].");
    }
}
```

Comment (na1307) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/UpdaterWorker.cs:45]: Ah, it was a problem with my test solution.

**Comment by josesimoes** [comment]
The `restore --force-evaluate` should indeed fix this. 
Can we please have this fix in production and get rid of the issue finally? It was reported in 2019... 😯

**Comment by mheguy** [comment]
Hey @jonjanego, would you be able to have someone from the team take a look at this to evaluate next steps?

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/LockFileUpdater.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    {
+        var lockPath = Path.Combine(Path.GetDirectoryName(projectPath), "packages.lock.json");
+        logger.Log($"  Running for lock file");
+        if (!File.Exists(lockPath))
+        {
+            logger.Log($"    File [{Path.GetRelativePath(repoRootPath, lockPath)}] does not exist.");
+            return;
+        }
+
+        var (exitCode, stdout, stderr) = await ProcessEx.RunAsync("dotnet", $"restore --force-evaluate {projectPath}");
```
Comment (brettfo) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/LockFileUpdater.cs:18]: Set the `workingDirectory` parameter here.

**Review by brettfo — APPROVED** [review]

**Review by abdulapopoola — APPROVED** [review]

**Comment by abdulapopoola** [comment]
thanks @na1307 for the contribution!

**Comment by sliekens** [comment]
Actually huge that this work is now completed

**Comment by ffried** [comment]
This seems to fail if the `csproj` has spaces in its name: 
```
2024-09-26T06:18:14.0831914Z Running for project file [VINMAPP File API Client.csproj]
2024-09-26T06:18:14.0833473Z Updating project [/home/dependabot/dependabot-updater/repo/VINMAPP File API Client.csproj]
2024-09-26T06:18:14.0834964Z   Running for SDK-style project
2024-09-26T06:18:14.0836255Z     Found incorrect [PackageReference] version attribute in [VINMAPP File API Client.csproj].
2024-09-26T06:18:14.0837461Z     Saved [VINMAPP File API Client.csproj].
2024-09-26T06:18:14.0838171Z     Updating lock file
2024-09-26T06:18:14.0839802Z Temporarily removing `global.json` from `/home/dependabot/dependabot-updater/repo` and retaining MSBuild SDK declarations.
2024-09-26T06:18:14.0841256Z       Lock file update failed.
2024-09-26T06:18:14.0842000Z STDOUT:
2024-09-26T06:18:14.0842720Z MSBUILD : error MSB1008: Only one project can be specified.
2024-09-26T06:18:14.0847533Z     Full command line: '/usr/local/dotnet/current/sdk/8.0.303/MSBuild.dll -maxcpucount -verbosity:m -nologo -target:Restore -property:RestoreForceEvaluate=true /home/dependabot/dependabot-updater/repo/VINMAPP File API Client.csproj'
```

The project path should be quoted in the execution to make sure it's passed completely.

**Comment by na1307** [comment]
https://github.com/dependabot/dependabot-core/blob/ddb9722dd9ed00daf54b5115ccfe033c6bb910b7/nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/LockFileUpdater.cs#L21

Oh, I think I forgot the quotation marks.

**Comment by ffried** [comment]
@na1307 I've forked the project and found more issues where paths are not quoted. I'm working on a fix.

**Comment by skolima** [comment]
Looks like (from the logs) that Dependabot now generates lock file updates - great! But sadly at the same time it started failing to create the PR for us, GitHub responds with 400 `{"errors":[{"status":400,"title":"Bad Request","detail":"The request contains invalid or unauthorized changes"}]}`.

<details><summary>log file excerpt</summary>
```
updater | 2024/09/27 10:21:59 INFO <job_892011036> Updating Aspire.Azure.Storage.Blobs from 8.2.0 to 8.2.1
updater | running NuGet updater:
/opt/nuget/NuGetUpdater/NuGetUpdater.Cli update --repo-root /home/dependabot/dependabot-updater/repo --solution-or-project /home/dependabot/dependabot-updater/repo/src/MyProject/MyProject.csproj --dependency Aspire.Azure.Storage.Blobs --new-version 8.2.1 --previous-version 8.2.0 --result-output-path /tmp/update-result.json --verbose
  proxy | 2024/09/27 10:22:02 [250] GET https://api.nuget.org:443/v3-flatcontainer/aspire.azure.storage.blobs/8.2.1/aspire.azure.storage.blobs.8.2.1.nupkg
  proxy | 2024/09/27 10:22:02 [250] 200 https://api.nuget.org:443/v3-flatcontainer/aspire.azure.storage.blobs/8.2.1/aspire.azure.storage.blobs.8.2.1.nupkg
  proxy | 2024/09/27 10:22:07 Posting metrics to remote API endpoint
  proxy | 2024/09/27 10:22:07 Successfully posted metrics data via api client
updater |   Updating [.config/dotnet-tools.json] file.
    Dependency [Aspire.Azure.Storage.Blobs] not found.
  No global.json file found.
Running for project file [src/MyProject/MyProject.csproj]
Updating project [/home/dependabot/dependabot-updater/repo/src/MyProject/MyProject.csproj]
  Running for SDK-style project
    Found incorrect [PackageVersion] version attribute in [src/Directory.Packages.props].
    Saved [src/Directory.Packages.props].
    Updating lock file
Update complete.
  Writing update result to [/tmp/update-result.json].
2024/09/27 10:22:12 INFO <job_892011036> update result: {
  "ErrorType": null,
  "ErrorDetails": null
}
updater | The contents of file [src/Directory.Packages.props] were updated.
updater | The contents of file [src/MyProject/packages.lock.json] were updated.
updater | 2024/09/27 10:22:12 INFO <job_892011036> Submitting Aspire.Azure.Storage.Blobs pull request for creation
  proxy | 2024/09/27 10:22:12 [252] GET https://api.github.com:443/repos/redacted/commits?per_page=100
  proxy | 2024/09/27 10:22:12 [252] 200 https://api.github.com:443/repos/redacted/commits?per_page=100
  proxy | 2024/09/27 10:22:12 [254] GET https://api.nuget.org:443/v3-flatcontainer/aspire.azure.storage.blobs/8.2.1/aspire.azure.storage.blobs.nuspec
2024/09/27 10:22:12 [254] 200 https://api.nuget.org:443/v3-flatcontainer/aspire.azure.storage.blobs/8.2.1/aspire.azure.storage.blobs.nuspec
  proxy | 2024/09/27 10:22:12 [256] GET https://api.github.com:443/repos/dotnet/aspire/releases?per_page=100
  proxy | 2024/09/27 10:22:12 [256] 200 https://api.github.com:443/repos/dotnet/aspire/releases?per_page=100
  proxy | 2024/09/27 10:22:12 [258] GET https://api.github.com:443/repos/dotnet/aspire/contents/
2024/09/27 10:22:12 [258] 200 https://api.github.com:443/repos/dotnet/aspire/contents/
  proxy | 2024/09/27 10:22:12 [260] GET https://api.github.com:443/repos/dotnet/aspire/contents/docs
  proxy | 2024/09/27 10:22:12 [260] 200 https://api.github.com:443/repos/dotnet/aspire/contents/docs
  proxy | 2024/09/27 10:22:12 [262] GET https://github.com:443/dotnet/aspire.git/info/refs?service=git-upload-pack
2024/09/27 10:22:12 [262] 200 https://github.com:443/dotnet/aspire.git/info/refs?service=git-upload-pack
  proxy | 2024/09/27 10:22:12 [264] GET https://api.github.com:443/repos/dotnet/aspire/contents/?ref=v8.2.1
2024/09/27 10:22:12 [264] 200 https://api.github.com:443/repos/dotnet/aspire/contents/?ref=v8.2.1
  proxy | 2024/09/27 10:22:12 [266] GET https://api.github.com:443/repos/dotnet/aspire/contents/docs?ref=v8.2.1
  proxy | 2024/09/27 10:22:12 [266] 200 https://api.github.com:443/repos/dotnet/aspire/contents/docs?ref=v8.2.1
  proxy | 2024/09/27 10:22:12 [268] GET https://github.com:443/dotnet/aspire.git/info/refs?service=git-upload-pack
  proxy | 2024/09/27 10:22:12 [268] 200 https://github.com:443/dotnet/aspire.git/info/refs?service=git-upload-pack
  proxy | 2024/09/27 10:22:13 [270] GET https://api.github.com:443/repos/dotnet/aspire/commits?sha=v8.2.0
2024/09/27 10:22:13 [270] 200 https://api.github.com:443/repos/dotnet/aspire/commits?sha=v8.2.0
  proxy | 2024/09/27 10:22:13 [272] GET https://api.github.com:443/repos/dotnet/aspire/commits?sha=v8.2.1
  proxy | 2024/09/27 10:22:13 [272] 200 https://api.github.com:443/repos/dotnet/aspire/commits?sha=v8.2.1
  proxy | 2024/09/27 10:22:13 [274] GET https://api.github.com:443/repos/dotnet/aspire/commits?sha=v8.2.0
  proxy | 2024/09/27 10:22:13 [274] 200 https://api.github.com:443/repos/dotnet/aspire/commits?sha=v8.2.0
  proxy | 2024/09/27 10:22:13 [276] GET https://api.github.com:443/repos/dotnet/aspire/commits?sha=v8.2.1
  proxy | 2024/09/27 10:22:13 [276] 200 https://api.github.com:443/repos/dotnet/aspire/commits?sha=v8.2.1
  proxy | 2024/09/27 10:22:13 [278] GET https://api.github.com:443/repos/dotnet/aspire/commits?sha=v8.2.0
  proxy | 2024/09/27 10:22:13 [278] 200 https://api.github.com:443/repos/dotnet/aspire/commits?sha=v8.2.0
  proxy | 2024/09/27 10:22:13 [280] GET https://api.github.com:443/repos/dotnet/aspire/commits?sha=v8.2.1
2024/09/27 10:22:13 [280] 200 https://api.github.com:443/repos/dotnet/aspire/commits?sha=v8.2.1
  proxy | 2024/09/27 10:22:13 [282] POST /update_jobs/892011036/create_pull_request
  proxy | 2024/09/27 10:22:13 [282] 400 /update_jobs/892011036/create_pull_request
  proxy | 2024/09/27 10:22:13 [284] POST /update_jobs/892011036/record_update_job_unknown_error
  proxy | 2024/09/27 10:22:13 [284] 204 /update_jobs/892011036/record_update_job_unknown_error
  proxy | 2024/09/27 10:22:13 [286] POST /update_jobs/892011036/record_update_job_error
  proxy | 2024/09/27 10:22:13 [286] 204 /update_jobs/892011036/record_update_job_error
  proxy | 2024/09/27 10:22:13 [288] POST /update_jobs/892011036/increment_metric
2024/09/27 10:22:13 [288] 204 /update_jobs/892011036/increment_metric
  proxy | 2024/09/27 10:22:13 [290] POST /update_jobs/892011036/record_update_job_unknown_error
2024/09/27 10:22:13 [290] 204 /update_jobs/892011036/record_update_job_unknown_error
updater | 2024/09/27 10:22:13 ERROR <job_892011036> Error processing Aspire.Azure.Storage.Blobs (Dependabot::ApiError)
2024/09/27 10:22:13 ERROR <job_892011036> {"errors":[{"status":400,"title":"Bad Request","detail":"The request contains invalid or unauthorized changes"}]}
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/lib/dependabot/api_client.rb:42:in `block in create_pull_request'
updater | 2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/opentelemetry-api-1.2.3/lib/opentelemetry/trace/tracer.rb:37:in `block in in_span'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/opentelemetry-api-1.2.3/lib/opentelemetry/trace.rb:70:in `block in with_span'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/opentelemetry-api-1.2.3/lib/opentelemetry/context.rb:87:in `with_value'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/opentelemetry-api-1.2.3/lib/opentelemetry/trace.rb:70:in `with_span'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/opentelemetry-api-1.2.3/lib/opentelemetry/trace/tracer.rb:37:in `in_span'
updater | 2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/lib/dependabot/api_client.rb:34:in `create_pull_request'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/call_validation_2_7.rb:734:in `bind_call'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/call_validation_2_7.rb:734:in `block in create_validator_procedure_fast2'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/lib/dependabot/service.rb:57:in `create_pull_request'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/call_validation_2_7.rb:734:in `bind_call'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/call_validation_2_7.rb:734:in `block in create_validator_procedure_fast2'
updater | 2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/lib/dependabot/updater/operations/update_all_versions.rb:325:in `create_pull_request'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/call_validation_2_7.rb:687:in `bind_call'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/call_validation_2_7.rb:687:in `block in create_validator_procedure_fast1'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/lib/dependabot/updater/operations/update_all_versions.rb:186:in `check_and_create_pull_request'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/call_validation_2_7.rb:687:in `bind_call'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/call_validation_2_7.rb:687:in `block in create_validator_procedure_fast1'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/lib/dependabot/updater/operations/update_all_versions.rb:94:in `check_and_create_pr_with_error_handling'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/call_validation_2_7.rb:687:in `bind_call'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/call_validation_2_7.rb:687:in `block in create_validator_procedure_fast1'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/lib/dependabot/updater/operations/update_all_versions.rb:59:in `block in perform'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/lib/dependabot/updater/operations/update_all_versions.rb:59:in `each'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/lib/dependabot/updater/operations/update_all_versions.rb:59:in `perform'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/call_validation.rb:270:in `bind_call'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/call_validation.rb:270:in `validate_call'
updater | 2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/_methods.rb:277:in `block in _on_method_added'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/lib/dependabot/updater/operations/group_update_all_versions.rb:137:in `block in run_ungrouped_dependency_updates'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/lib/dependabot/updater/operations/group_update_all_versions.rb:122:in `each'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/lib/dependabot/updater/operations/group_update_all_versions.rb:122:in `run_ungrouped_dependency_updates'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/call_validation.rb:270:in `bind_call'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/call_validation.rb:270:in `validate_call'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/_methods.rb:277:in `block in _on_method_added'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/lib/dependabot/updater/operations/group_update_all_versions.rb:65:in `perform'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/call_validation.rb:270:in `bind_call'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/call_validation.rb:270:in `validate_call'
updater | 2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11444/lib/types/private/methods/_methods.rb:277:in `block in _on_method_added'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/lib/dependabot/updater.rb:45:in `run'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/lib/dependabot/update_files_command.rb:44:in `block in perform_job'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/opentelemetry-api-1.2.3/lib/opentelemetry/trace/tracer.rb:37:in `block in in_span'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/opentelemetry-api-1.2.3/lib/opentelemetry/trace.rb:70:in `block in with_span'
updater | 2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/opentelemetry-api-1.2.3/lib/opentelemetry/context.rb:87:in `with_value'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/opentelemetry-api-1.2.3/lib/opentelemetry/trace.rb:70:in `with_span'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/opentelemetry-api-1.2.3/lib/opentelemetry/trace/tracer.rb:37:in `in_span'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/lib/dependabot/update_files_command.rb:18:in `perform_job'
2024/09/27 10:22:13 ERROR <job_892011036> /home/dependabot/dependabot-updater/lib/dependabot/base_command.rb:37:in `run'
2024/09/27 10:22:13 ERROR <job_892011036> bin/update_files.rb:46:in `<main>'
  proxy | 2024/09/27 10:22:13 [292] PATCH /update_jobs/892011036/mark_as_processed
  proxy | 2024/09/27 10:22:13 [292] 204 /update_jobs/892011036/mark_as_processed
updater | 2024/09/27 10:22:13 INFO <job_892011036> Finished job processing
updater | 2024/09/27 10:22:13 INFO Results:
Dependabot encountered '8' error(s) during execution, please check the logs for more details.
+-------------------------------------------------------------+
|                Dependencies failed to update                |
+---------------------------------------+---------------------+
| Aspire.Npgsql                         | unknown_error       |
| xunit                                 | unknown_error       |
| Azure.Storage.Blobs                   | unknown_error       |
| Microsoft.Extensions.ServiceDiscovery | update_not_possible |
| Azure.Identity                        | unknown_error       |
| Aspire.Azure.Messaging.EventHubs      | unknown_error       |
| Scrutor                               | unknown_error       |
| Aspire.Azure.Storage.Blobs            | unknown_error       |
+---------------------------------------+---------------------+
Failure running container b884a6be21bc651d8906a802f930c26301b1660c800a76d1c530d059922d3175
Cleaned up container b884a6be21bc651d8906a802f930c26301b1660c800a76d1c530d059922d3175
  proxy | 2024/09/27 10:22:14 111/1139 calls cached (9%)
  proxy | 2024/09/27 10:22:14 Posting metrics to remote API endpoint
  proxy | 2024/09/27 10:22:14 Successfully posted metrics data via api client
```
</details>

**Comment by rkm** [comment]
> Looks like (from the logs) that Dependabot now generates lock file updates - great! But sadly at the same time it started failing to create the PR for us, GitHub responds with 400 `{"errors":[{"status":400,"title":"Bad Request","detail":"The request contains invalid or unauthorized changes"}]}`.
> 
> log file excerpt

Is there something to fix here, or is it just a case of waiting until the app is updated?

**Comment by skolima** [comment]
@rkm dependabot-core had a release yesterday, so I assume the "fixes" are in. They have not resolved the issue for me. I'll try to prepare a smaller repro.

**Comment by skolima** [comment]
Created a new issue with repro steps: https://github.com/dependabot/dependabot-core/issues/10730

**Comment by baynezy** [comment]
@skolima - I am getting the same problem

> Looks like (from the logs) that Dependabot now generates lock file updates - great! But sadly at the same time it started failing to create the PR for us, GitHub responds with 400 {"errors":[{"status":400,"title":"Bad Request","detail":"The request contains invalid or unauthorized changes"}]}.

I have looked into it and the issue I see is that if the dependency update causes a dependency downgrade, it will break the whole run of dependabot. Previously it would update the version in csproj only, which would then show as an issue when restoring in your CI pipeline. Now so it can update the package.lock.json files it needs to do a build; which fails because of the package downgrade. So it cannot create the PR. 

This is worse because it breaks the whole run and won't create any NuGet-related PRs.

**Comment by vancodocton** [comment]
It worked partially in my case.  Dependabot does great work for my projects in /src folder. But for my test projects in /test that reference project in src, it does not bump their lock files. I need to restore and push new commit to pass my CI

**Comment by vmcbaptista** [comment]
Does anyone know if the issue I raised here was addressed? https://github.com/orgs/community/discussions/144979
